// @sap/approuter/lib/passport/oauth-configuration.js
'use strict';

var _ = require('lodash');
var OAuth2Strategy = require('./oauth2-strategy');
var BasicOAuth2Strategy = require('./basic-oauth2-strategy');
var utils = require('./utils');

module.exports = {
  getUserSerializer: function (user, callback) {
    if (user && user.name) {
      callback(null, JSON.stringify(user));
    } else {
      callback('No user name found');
    }
  },

  getUserDeserializer: function (sUser, callback) {
    if (sUser === '') {
      callback(null, {});
    } else {
      var user = JSON.parse(sUser);
      if (user) {
        callback(null, user);
      } else {
        callback('User not found ' + sUser);
      }
    }
  },

  getXSUAAOauthStrategy: function (req, cb) {
    createStrategy(req, function(err, options, doneCallback) {
      if (err) { return cb(err); }
      cb(null, new OAuth2Strategy(options, doneCallback));
    });
  },

  getBasicOauthStrategy: function (req, credentials, cb) {
    createStrategy(req, function(err, options, doneCallback) {
      if (err) { return cb(err); }
      cb(null, new BasicOAuth2Strategy(_.assign(options, credentials), doneCallback));
    });
  }
};

function createStrategy(req, cb) {
  utils.loadOauthOptions(req, function(err, options) {
    if (err) { return cb(err); }

    function authenticationDone(req, params, done) {
      var tokenContext = {
        accessToken: params['access_token'],
        refreshToken: params['refresh_token'],
        expiresIn: params['expires_in'],
        scope: params.scope,
        oauthOptions: options
      };

      utils.storeToken(req, tokenContext, function(err) {
        if (!err) {
          req.app.approuter.emit('login', req.session);
        }
        done.apply(null, arguments);
      });
    }

    cb(null, options, authenticationDone);
  });
}
