// @sap/xsodata/lib/sql/createDeleteStatements.js
'use strict';

//Include
var sql = require('./sqlStatement');
var sqlTools = require('./sqlTools');

//Code
exports.createDeleteStatements = function (context, asyncDone) {
    var dbSegLast = context.oData.dbSegmentLast;

    context.logger.silly('sqlCollector', 'createDeleteStatements');

    var sqlContext = {
        context: context,
        netId: context.uniqueNetworkRequestID,
        reqId: context.uniqueRequestID,
        rIdOld: 1,
        dbSegLast: dbSegLast,
        systemQueryParameters: context.oData.systemQueryParameters
    };

    context.sql = sqlContext;

    dbSegLast.sql.stmContainer = new sql.DeleteContainer();

    startDeleteMasterDbSeg(sqlContext);
    return asyncDone(null, context);
};

/**
 * Create SQL statements for the master table (=end of resource path)
 * @param sqlContext
 */
function startDeleteMasterDbSeg(sqlContext) {
    var dbSeg = sqlContext.dbSegLast,
        stmtContainer = dbSeg.sql.stmContainer;

    sqlContext.context.sql.container = stmtContainer;

    sqlContext.context.logger.debug('sqlCollector', 'startPutMasterDbSeg');

    // Must be renamed rIdOld (from rId), in order to land in 'beforeTableName' of exitProcessor
    dbSeg.sql.rIdOld = sqlTools.rIdToOldTableName(dbSeg._Alias,sqlContext.netId, sqlContext.reqId, sqlContext.rIdOld);
    sqlContext.rIdOld++;

    //create statement for master table
    dbSeg.sql.stmContainer.createTmpDel = masterTableCreate(sqlContext, dbSeg.sql.rIdOld);

    //create insert statement for master table
    dbSeg.sql.stmContainer.insertTmpDel = masterTableInsertToDel(sqlContext, dbSeg.sql.rIdOld);

    //create select statement for master table
    dbSeg.sql.stmContainer.delete = masterTableDelete(sqlContext);

    if (sqlContext.context.db.isExternalHandledConnection === true) {
        // We only truncate and delete temp tables when db connection is handled external
        // otherwise we will self disconnect and temp tables will be deleted automatically

        // build the truncate statement for created temporary table;
        stmtContainer.createTmpDelTruncate = sql.buildTableStatement(sql.Truncate, sqlContext,
            stmtContainer.createTmpDel.table);
        sqlContext.context.networkContext.cleanSessionTruncateContainer.push(stmtContainer.createTmpDelTruncate);

        // build the drop statement for created temporary table;
        stmtContainer.createTmpDelDrop = sql.buildTableStatement(sql.Drop, sqlContext,
            stmtContainer.createTmpDel.table);
        sqlContext.context.networkContext.cleanSessionDropContainer.push(stmtContainer.createTmpDelDrop);
    }

}


function masterTableCreate(sqlContext, rIdOld) {
    sqlContext.context.logger.debug('sqlCollector', 'masterTableCreate');
    var dbSeg = sqlContext.dbSegLast;

    var stmCreate = new sql.Create();
    stmCreate.setModifiers(['local', 'temporary', dbSeg.entityType.tableStoreType]);
    stmCreate.setTableName(rIdOld);
    stmCreate.setAs(createSelectTemplate(dbSeg));
    stmCreate.setPostModifiers(['with no data']);
    return stmCreate;


    /**
     * Create SQL Statement 'Create Master'
     * @param dbSeg
     */
    function createSelectTemplate(dbSeg) {
        var stm = new sql.Select();

        //SELECT
        stm.addSelects(dbSeg.getQKeyProperties());
        if (dbSeg._ExpandedNavigations.length > 0) {
            stm.addSelects(dbSeg.getNonKeyNonSelectedProperties4Expansion());
        }
        stm.addSelects(dbSeg.getQNonKeyProperties());

        //FROM
        stm.setFrom(dbSeg.getAliasedTableName());
        return stm;
    }
}

function masterTableInsertToDel(sqlContext, rIdOld) {
    sqlContext.context.logger.debug('sqlCollector', 'masterTableInsertToNew');
    var dbSeg = sqlContext.dbSegLast;

    var subSelect = new sql.Select();
    subSelect.addSelects(dbSeg.getQKeyPropertiesForSelect());
    subSelect.addSelects(dbSeg.getQNonKeyPropertiesForSelect());
    subSelect.setFrom(dbSeg.getAliasedTableName());
    subSelect.addWhereKeyValuePairs(dbSeg.getQKeyWithValuesDB());

    var stm = new sql.Insert();
    stm.setTableName({ table: rIdOld });
    stm.setSubSelect(subSelect);

    return stm;
}


/**
 * Create select SQL statement for master table
 * @param sqlContext
 * @returns {exports.Select}
 */
function masterTableDelete(sqlContext) {
    sqlContext.context.logger.debug('sqlCollector', 'masterTableDelete');
    var dbSeg = sqlContext.dbSegLast;

    var delete1 = new sql.Delete();
    delete1.setFrom(dbSeg.getAliasedTableName());
    delete1.addWhereKeyValuePairs(dbSeg.getQKeyWithValues());

    return delete1;
}
