// @sap/xsodata/lib/sql/dataCollectorPut.js
'use strict';

//Include
var dataCollector = require('./dataCollector');
var sqlStatement = require('./sqlStatement');

exports.movePayloadFromSegmentToSelectStm = function (context, asyncDone) {
    context.logger.silly('dataCollector', 'movePayloadFromSegmentToSelectStm');
    //get dbSegment for inserting the payload
    var dbSeg = context.oData.dbSegmentLast;
    var stm = dbSeg.sql.stmContainer.insertTmp;
    if (!stm) {
        return asyncDone("Internal server error", context);
    }

    var max = dbSeg._recordNV.length;
    var i;
    for (i = 0; i < max; i++) {
        var nv = dbSeg._recordNV[i];
        stm.setValue(nv.name, nv.value);
    }
    return asyncDone(null, context);
};


exports.createTmpTable = function (context, asyncDone) {
    var dbSeg = context.oData.dbSegmentLast;
    var stms = [dbSeg.sql.stmContainer.createTmpOld, dbSeg.sql.stmContainer.createTmp];
    return dataCollector.execParallelNoResult(stms, context, asyncDone);
};


exports.insertOldDataToOldTable = function (context, asyncDone) {
    var dbSeg = context.oData.dbSegmentLast;
    return dataCollector.execParallelNoResultwithParam([dbSeg.sql.stmContainer.insertTmpOld], context, asyncDone);
};


exports.insertPayloadIntoTempTable = function (context, asyncDone) {
    context.logger.silly('dataCollectorPut', 'insertPayloadIntoTempTable');
    var client = context.db.client;
    var dbSeg = context.oData.dbSegmentLast;
    var stm = dbSeg.sql.stmContainer.insertTmp;

    var p = [];
    var sql = stm.toSqlHana(new sqlStatement.SqlBuildHanaContext(context), p);
    context.logger.debug('SQL Exec', 'SQL: \n' + sql);
    client.prepare(sql, function (err, statement) {
        if (err) {
            context.logger.info('SQL Exec', 'Error: \n' + err);
            context.logger.info('SQL Exec', 'SQL: \n' + sql);
            return asyncDone(err, context);
        }

        context.logger.debug('SQL Exec', 'Parameters:\n' + JSON.stringify(p));
        statement.exec(p, function (err) {
            if (err) {
                context.logger.info('SQL Exec', 'Error: \n' + err);
                context.logger.info('SQL Exec', 'SQL: \n' + sql);
                context.logger.info('SQL Exec', 'Parameters:\n' + JSON.stringify(p));
                return asyncDone(err, context);
            }
            return asyncDone(null, context);
        });
    });
};


exports.insertTmpTableToRealTable = function (context, asyncDone) {
    context.logger.silly('dataCollectorPut', 'insertTmpTableToRealTable');
    var dbSeg = context.oData.dbSegmentLast;
    return dataCollector.execParallelNoResultwithParam([dbSeg.sql.stmContainer.updateReal], context,
        dataCollector.getAffectedRowsValidator(1, asyncDone),
        /**
         * true ->  provide affectedRows parameter.
         */
        true);
};

exports.commit = function (context, asyncDone) {
    context.logger.silly('dataCollectorPut', 'commit');
    if (context.batchContext) {
        //When running in batch the commit is performed by the batch executor
        return asyncDone(null, context);
    }

    var client = context.db.client;
    client.commit(function (err) {
        if (err) {
            context.logger.info('SQL Exec', 'Commit Error: \n' + err);
            return asyncDone(err, context);
        }
        return asyncDone(null, context);
    });
};

/**
 * Executes truncation of temporary created tables
 *
 * @param {Object} context The xsodata context
 * @param {Function} asyncDone async waterfall callback
 */
exports.truncateTempTables = function (context, asyncDone) {
    var statements;
    context.logger.silly('dataCollectorPut', 'truncateTempTables');

    statements = [
        context.sql.container.createTmpTruncate,
        context.sql.container.createTmpOldTruncate
    ];

    return dataCollector.execTempTableStatements(statements, context, asyncDone);
};

/**
 * Executes deletion of temporary created tables
 *
 * @param {Object} context The xsodata context
 * @param {Function} asyncDone async waterfall callback
 */
exports.dropTempTables = function (context, asyncDone) {
    var statements;
    context.logger.silly('dataCollectorPut', 'dropTempTables');

    statements = [
        context.sql.container.createTmpDrop,
        context.sql.container.createTmpOldDrop
    ];

    return dataCollector.execTempTableStatements(statements, context, asyncDone);
};
