// @sap/audit-logging/lib/messages/MessageBase.js
'use strict';

var currentTransport = require('../transports/current-transport');
var moment = require('moment');
var utils = require('../utils');


module.exports = MessageBase;

function MessageBase(action) {
  this._verb = { action: action, timestamp: moment.utc().format('YYYY-MM-DDTHH:mm:ss.SSSZ') };
  this._custom = {};
}

MessageBase.prototype.customAttribute = function (name, value) {
  utils.validate.isValidString(name, 'Custom attribute name');
  utils.validate.isProvided(value, 'Custom attribute value');

  this._custom[name] = value;
  return this;
};

MessageBase.prototype.by = function (driverOfAction) {
  return this.customAttribute('driver of action', driverOfAction);
};

MessageBase.prototype.accessChannel = function (accessChannel) {
  return this.customAttribute('access channel', accessChannel);
};

MessageBase.prototype.externalIP = function (externalIP) {
  return this.customAttribute('external IP', externalIP);
};

MessageBase.prototype.category = function (category) {
  return this.customAttribute('category', category);
};

MessageBase.prototype._stringifyParts = function () {
  var parts = {};
  parts.verb = 'verb=' + utils.stringify(this._verb);
  if (utils.hasContent(this._custom)) {
    parts.custom = 'custom=' + utils.stringify(this._custom);
  }
  return parts;
};

MessageBase.prototype._outputToTransport = function (callback) {
  utils.validate.callback(callback);
  currentTransport.log(this, callback);
};
